File: src/main/java/me/redepicness/bingo/command/BingoWorldsCommandArgument.java
package me.redepicness.bingo.command;

import cloud.commandframework.arguments.CommandArgument;
import cloud.commandframework.arguments.parser.ArgumentParseResult;
import cloud.commandframework.arguments.parser.ArgumentParser;
import cloud.commandframework.context.CommandContext;
import cloud.commandframework.exceptions.parsing.NoInputProvidedException;
import io.leangen.geantyref.TypeToken;
import me.redepicness.bingo.AnotherBingoPlugin;
import me.redepicness.bingo.world.BingoWorlds;
import org.bukkit.NamespacedKey;
import org.bukkit.command.CommandSender;
import org.bukkit.plugin.java.JavaPlugin;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Optional;
import java.util.Queue;

public class BingoWorldsCommandArgument extends CommandArgument<CommandSender, BingoWorlds> {

    public static @NonNull BingoWorldsCommandArgument of(final @NonNull String name) {
        return new BingoWorldsCommandArgument(true, name);
    }

    public static @NonNull BingoWorldsCommandArgument optional(final @NonNull String name) {
        return new BingoWorldsCommandArgument(false, name);
    }

    public BingoWorldsCommandArgument(boolean required, @NonNull String name) {
        super(
                required,
                name,
                new BingoWorldsParser(),
                "",
                TypeToken.get(BingoWorlds.class),
                null
        );
    }

    public static final class BingoWorldsParser implements ArgumentParser<CommandSender, BingoWorlds> {

        @Override
        public @NonNull ArgumentParseResult<@NonNull BingoWorlds> parse(
                @NonNull CommandContext<@NonNull CommandSender> context,
                @NonNull Queue<@NonNull String> inputQueue
        ) {
            String input = inputQueue.peek();
            if (input == null) {
                return ArgumentParseResult.failure(new NoInputProvidedException(BingoWorldsParser.class, context));
            }

            NamespacedKey key;
            try {
                key = new NamespacedKey("abp", "bingo-"+input);
            }
            catch (Exception e) {
                return ArgumentParseResult.failure(new IllegalArgumentException("Could not parse " + getCurrentArgumentInfo(context) + ", invalid key! (" + input + ")"));
            }
            Optional<BingoWorlds> worlds = JavaPlugin.getPlugin(AnotherBingoPlugin.class).getWorldManager().getWorlds(key);
            if (worlds.isEmpty()) {
                return ArgumentParseResult.failure(new IllegalArgumentException("Invalid key " + getCurrentArgumentInfo(context) + ", no bingoWorlds found! (" + input + ")"));
            }

            inputQueue.poll();
            return ArgumentParseResult.success(worlds.get());
        }

        @Override
        public @NonNull List<@NonNull String> suggestions(@NonNull CommandContext<CommandSender> commandContext, @NonNull String input) {
            return JavaPlugin.getPlugin(AnotherBingoPlugin.class).getWorldManager().getAllWorlds()
                    .stream()
                    .map(w -> w.getOverworld().getKey().getKey().substring(6))
                    .filter(wk -> wk.startsWith(input))
                    .toList();
        }

        @Override
        public boolean isContextFree() {
            return true;
        }

    }

    private static @NotNull String getCurrentArgumentInfo(@NonNull CommandContext<@NonNull CommandSender> commandContext) {
        CommandArgument<@NonNull CommandSender, ?> currentArgument = commandContext.getCurrentArgument();
        if (currentArgument == null) return "unknown";
        return  "BingoWorlds '" + currentArgument.getName() + "'";
    }

}

File: src/main/kotlin/me/redepicness/bingo/command/argument/BingoWorldsCommandArgument.kt
package me.redepicness.bingo.command.argument

import cloud.commandframework.arguments.CommandArgument
import cloud.commandframework.arguments.parser.ArgumentParseResult
import cloud.commandframework.arguments.parser.ArgumentParser
import cloud.commandframework.context.CommandContext
import cloud.commandframework.exceptions.parsing.NoInputProvidedException
import io.leangen.geantyref.TypeToken
import me.redepicness.bingo.world.BingoWorlds
import me.redepicness.bingo.world.WorldManager
import org.bukkit.NamespacedKey
import org.bukkit.command.CommandSender
import java.util.*

class BingoWorldsCommandArgument(required: Boolean, name: String) : CommandArgument<CommandSender, BingoWorlds>(
        required,
        name,
        BingoWorldsParser(),
        "",
        TypeToken.get(BingoWorlds::class.java),
        null
) {
    class BingoWorldsParser : ArgumentParser<CommandSender, BingoWorlds> {
        override fun parse(context: CommandContext<CommandSender>, inputQueue: Queue<String>): ArgumentParseResult<BingoWorlds> {
            val input = inputQueue.peek() ?: return ArgumentParseResult.failure(NoInputProvidedException(BingoWorldsParser::class.java, context))
            val key = try {
                NamespacedKey("abp", "bingo-$input")
            } catch (e: Exception) {
                return ArgumentParseResult.failure(IllegalArgumentException("Could not parse " + getCurrentArgumentInfo(context) + ", invalid key! (" + input + ")"))
            }
            val worlds = WorldManager.getWorlds(key)
                ?: return ArgumentParseResult.failure(IllegalArgumentException("Invalid key " + getCurrentArgumentInfo(context) + ", no bingoWorlds found! (" + input + ")"))
            inputQueue.poll()
            return ArgumentParseResult.success(worlds)
        }

        override fun suggestions(commandContext: CommandContext<CommandSender>, input: String): List<String> {
            return WorldManager.allWorlds
                    .stream()
                    .map { it.overworld.key.key.substring(6) }
                    .filter { it.startsWith(input) }
                    .toList()
        }

        override fun isContextFree(): Boolean {
            return true
        }
    }

    companion object {
        fun of(name: String): BingoWorldsCommandArgument {
            return BingoWorldsCommandArgument(true, name)
        }

        fun optional(name: String): BingoWorldsCommandArgument {
            return BingoWorldsCommandArgument(false, name)
        }

        private fun getCurrentArgumentInfo(commandContext: CommandContext<CommandSender>): String {
            val currentArgument = commandContext.currentArgument ?: return "unknown"
            return "BingoWorlds '" + currentArgument.name + "'"
        }
    }
}
